### PR TITLE
Add Android support @NonNull/@Nullable annotations

### PR DESCRIPTION
**Pull request by svenjacobs**
This implements #165

### CODE REVIEW

**Comment by orhanobut** [comment]
Definitely great job! I'll make a thorough review and come back to you. 👍 I wanted to apply support-annotation for a long time.

File: logger/src/main/java/com/orhanobut/logger/AndroidLogAdapter.java [hunk 1]
```diff
@@ -1,23 +1,28 @@
...
+    this.formatStrategy = checkNotNull(formatStrategy);
   }
 
-  @Override public boolean isLoggable(int priority, String tag) {
+  @Override public boolean isLoggable(int priority, @Nullable String tag) {
     return true;
   }
 
-  @Override public void log(int priority, String tag, String message) {
+  @Override public void log(int priority, @Nullable String tag, @NonNull String message) {
```
Comment (orhanobut) [inline logger/src/main/java/com/orhanobut/logger/AndroidLogAdapter.java]: `tag` parameter should be `@NonNull`

File: logger/src/main/java/com/orhanobut/logger/CsvFormatStrategy.java [hunk 2]
```diff
@@ -90,26 +100,31 @@ private String formatTag(String tag) {
...
-    public Builder logStrategy(LogStrategy val) {
+    @NonNull
+    public Builder logStrategy(@Nullable LogStrategy val) {
       logStrategy = val;
       return this;
     }
 
-    public Builder tag(String tag) {
+    @NonNull
+    public Builder tag(@Nullable String tag) {
```
Comment (orhanobut) [inline logger/src/main/java/com/orhanobut/logger/CsvFormatStrategy.java]: `tag` parameter should be `@NonNull`

File: logger/src/main/java/com/orhanobut/logger/CsvFormatStrategy.java [hunk 3]
```diff
@@ -20,23 +24,28 @@
...
   private static final String SEPARATOR = ",";
 
-  private final Date date;
-  private final SimpleDateFormat dateFormat;
-  private final LogStrategy logStrategy;
-  private final String tag;
+  @NonNull private final Date date;
+  @NonNull private final SimpleDateFormat dateFormat;
+  @NonNull private final LogStrategy logStrategy;
+  @Nullable private final String tag;
```
Comment (orhanobut) [inline logger/src/main/java/com/orhanobut/logger/CsvFormatStrategy.java]: `tag` field should be `@NonNull`

File: logger/src/main/java/com/orhanobut/logger/DiskLogAdapter.java [hunk 4]
```diff
@@ -1,22 +1,27 @@
...
+    this.formatStrategy = checkNotNull(formatStrategy);
   }
 
-  @Override public boolean isLoggable(int priority, String tag) {
+  @Override public boolean isLoggable(int priority, @Nullable String tag) {
     return true;
   }
 
-  @Override public void log(int priority, String tag, String message) {
+  @Override public void log(int priority, String tag, @NonNull String message) {
```
Comment (orhanobut) [inline logger/src/main/java/com/orhanobut/logger/DiskLogAdapter.java]: There is missing annotation for `tag` parameter which should be `@NonNull`

File: logger/src/main/java/com/orhanobut/logger/DiskLogStrategy.java [hunk 5]
```diff
@@ -3,41 +3,47 @@
...
+  @NonNull private final Handler handler;
 
-  public DiskLogStrategy(Handler handler) {
-    this.handler = handler;
+  public DiskLogStrategy(@NonNull Handler handler) {
+    this.handler = checkNotNull(handler);
   }
 
-  @Override public void log(int level, String tag, String message) {
+  @Override public void log(int level, @Nullable String tag, @NonNull String message) {
```
Comment (orhanobut) [inline logger/src/main/java/com/orhanobut/logger/DiskLogStrategy.java]: `tag` parameter should be `@NonNull`

File: logger/src/main/java/com/orhanobut/logger/LogAdapter.java [hunk 6]
```diff
@@ -1,8 +1,11 @@
...
+
 public interface LogAdapter {
 
-  boolean isLoggable(int priority, String tag);
+  boolean isLoggable(int priority, @Nullable String tag);
 
-  void log(int priority, String tag, String message);
-}
\ No newline at end of file
+  void log(int priority, @Nullable String tag, @NonNull String message);
```
Comment (orhanobut) [inline logger/src/main/java/com/orhanobut/logger/LogAdapter.java]: `tag` parameter should be `@NonNull`

File: logger/src/main/java/com/orhanobut/logger/PrettyFormatStrategy.java [hunk 7]
```diff
@@ -169,36 +188,42 @@ private String formatTag(String tag) {
...
-    public Builder logStrategy(LogStrategy val) {
+    @NonNull
+    public Builder logStrategy(@Nullable LogStrategy val) {
       logStrategy = val;
       return this;
     }
 
-    public Builder tag(String tag) {
+    @NonNull
+    public Builder tag(@Nullable String tag) {
```
[the remark this answers is not in the document]
Comment (svenjacobs) [reply logger/src/main/java/com/orhanobut/logger/PrettyFormatStrategy.java]: The point is that in builders the intermediate properties can be `null` temporarily until they have been set by a setter method. Also some `build` methods check for `null` and will apply a default value if property is `null`.

**Comment by svenjacobs** [comment]
Regarding `@NonNull` for `tag`, are you sure? For instance `formatTag()` in `PrettyFormatStrategy` explicitely checks for `null` values of `tag` and applies a default value. Other usages of `tag` also seem to be null-aware. That's why I applied `@Nullable` to `tag`.

**Comment by orhanobut** [comment]
In regarding to `FormatStrategy`, yes it should be actually `@Nullable` but I believe other places I mentioned should not be null. The reason is that passing 'null' actually breaks logging, underlying system ignores the log.

File: logger/src/main/java/com/orhanobut/logger/PrettyFormatStrategy.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (orhanobut) [reply logger/src/main/java/com/orhanobut/logger/PrettyFormatStrategy.java]: But this annotation applies for the given parameter, right? Unless you refer the field. In that case yes, we can leave the field `@Nullable`, makes sense.

File: logger/build.gradle [hunk 8]
```diff
@@ -19,11 +19,12 @@ android {
 }
 
 dependencies {
+  compile deps.supportAnnotations
```
[the remark this answers is not in the document]
Comment (svenjacobs) [reply logger/build.gradle]: Consider it done ;)

**Comment by svenjacobs** [comment]
Due to the absence of nullability annotations I wasn't sure regarding `tag` and was assuming from what the code told me ;) When adding `@NonNull` with a check of `checkNotNull()` this could potentially break clients that pass `null` to `tag` so I went the safe way.

**Comment by orhanobut** [comment]
Hmm, I am thinking that maybe we allow null and keep the changes for tag. We can apply an empty/null check in 'LogcatLogStrategy' . If the tag is empty/null we can put a default value in here. Because this is the only place that could break.

File: logger/src/main/java/com/orhanobut/logger/PrettyFormatStrategy.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (orhanobut) [reply logger/src/main/java/com/orhanobut/logger/PrettyFormatStrategy.java]: Let's keep it `@Nullable` to be in the safe side

**Comment by svenjacobs** [comment]
That sounds like a reasonable compromise :)

File: logger/src/test/java/com.orhanobut.logger/CsvFormatStrategyTest.java [hunk 9]
```diff
@@ -36,12 +39,12 @@
     FormatStrategy formatStrategy = CsvFormatStrategy.newBuilder()
         .tag("custom")
         .logStrategy(new LogStrategy() {
-          @Override public void log(int priority, String tag, String message) {
+          @Override public void log(int priority, String tag, @NonNull String message) {
```
Comment (orhanobut) [inline logger/src/test/java/com.orhanobut.logger/CsvFormatStrategyTest.java]: There is missing annotation for `tag` parameter. `@Nullable`

**Comment by svenjacobs** [comment]
After upgrading the project Travis seems to have a problem

```
You have not accepted the license agreements of the following SDK components:
  [Android SDK Platform 27, Android SDK Build-Tools 27.0.3].
  Before building your project, you need to accept the license agreements and complete the installation of the missing components using the Android Studio SDK Manager.
  Alternatively, to learn how to transfer the license agreements from one workstation to another, go to http://d.android.com/r/studio-ui/export-licenses.html
```

Can you please take a look @orhanobut?

File: logger/src/main/java/com/orhanobut/logger/LogcatLogStrategy.java [hunk 10]
```diff
@@ -1,11 +1,17 @@
...
 import android.util.Log;
 
+import static com.orhanobut.logger.Utils.checkNotNull;
+
 public class LogcatLogStrategy implements LogStrategy {
 
-  @Override public void log(int priority, String tag, String message) {
+  @Override public void log(int priority, @Nullable String tag, @NonNull String message) {
+    checkNotNull(message);
+
```
Comment (orhanobut) [inline logger/src/main/java/com/orhanobut/logger/LogcatLogStrategy.java]: In here, we can check empty/null case of tag and provide a default tag. "NO_TAG" would be ok.

**Comment by orhanobut** [comment]
I'll check it out Travis now

File: logger/src/main/java/com/orhanobut/logger/LogcatLogStrategy.java (hunk 10 above)
Comment (orhanobut) [reply logger/src/main/java/com/orhanobut/logger/LogcatLogStrategy.java]: That'd be great if you also add a test to cover this case by confirming that we use "NO_TAG"

**Comment by orhanobut** [comment]
Travis is missing corresponding changes. If you make the changes in `.travis.yml` file, everything should be fine. Although I'm not sure if travis supports latest versions, if not we have to downgrade them unfortunately to the version Travis supports

```
 - build-tools-27.0.3
 - android-27
```

File: logger/src/main/java/com/orhanobut/logger/LogcatLogStrategy.java (hunk 10 above)
Comment (svenjacobs) [reply logger/src/main/java/com/orhanobut/logger/LogcatLogStrategy.java]: Done

**Comment by orhanobut** [comment]
I'll merge the PR and fix Travis. Thanks for pull request 👍

**Pull request merged**